src/correction.cc

```diff
@@ -334,19 +334,71 @@ Binning::Binning(const JSONObject& json, const Correction& context)
     flow_ = _FlowBehavior::value;
     default_value = resolve_content(flowbehavior, context);
   }
-  bins_.reserve(edges.size());
-  // first bin is never accessed for content in range (corresponds to std::upper_bound underflow)
-  // use it to store default value
-  bins_.push_back({*edges.begin(), std::move(default_value)});
-  for (size_t i=0; i < content.Size(); ++i) {
-    bins_.push_back({edges[i + 1], resolve_content(content[i], context)});
+
+  // set bin contents
+  contents_.push_back(std::move(default_value));
+  for (size_t i=0; i < content.Size(); ++i)
+    contents_.push_back(resolve_content(content[i], context));
+
+  // helper function to check whether two floating point values are equal within a relative tolerance, a la math.isclose
+  // always returns false if a or b are a positive or negative infinity.
+  auto is_close = [](double a, double b) {
+    if (std::isinf(a) || std::isinf(b))
+        return false;
+    const double relative_tolerance = 1e-6;
+    return std::fabs(b - a) <= relative_tolerance * std::max(std::fabs(a), std::fabs(b));
+  };
+
+  // check whether we have uniform binning
+  bool isBinningUniform = true;
+  const double firstBinWidth = edges.size() > 1 ? edges[1] - edges[0] : 0.;
+  for (auto i = 1u; i < edges.size() - 1; ++i) {
```

Comment:
I was inclined to favor checking exact equality, i.e. that
```cpp
auto as_uniform = _UniformBins{/*lower*/edges.front(), /*upper*/edges.back(), /*n*/edges.size() - 1u};
for (auto i = 1u; i < edges.size() - 1; ++i) {
  if ( as_uniform.n * (edges[i] - as_uniform.lower) / (as_uniform.upper - as_uniform.lower) != (double) i ) {
    isBinningUniform = false;
    break;
  }
}
```
Here I'm using what our eventual lookup is, as opposed to `edges[i+1] - edges[i] == edges[1] - edges[0]` to be more sure the bin lookup is as close as possible to the behavior we would get by doing `std::upper_bound`. But that means whether or not we flag a binning as uniform is going to be very fickle. Here are 3 plausible ways of creating a "uniform" binning:
```python
import numpy as np
n, lo, up = 11, 0.1, 1.3
a = np.arange(n+1)*((up-lo)/n) + lo
b = np.arange(n+1)*(up-lo)/n + lo
c = np.arange(lo, up + np.finfo("d").eps, (up-lo)/n)
```
and `np.sum((a==c)&(b==c)&(a==b))` is 4, i.e. only 4 of our 12 edges actually coincide exactly in all three methods 😢 
Actually, none of those 3 methods are invertible:
```pythone
>>> n*(a - lo)/(up - lo) == np.arange(n+1)
array([ True, False,  True,  True, False,  True,  True, False,  True,
       False,  True,  True])
>>> n*(b - lo)/(up - lo) == np.arange(n+1)
array([ True, False,  True,  True, False,  True, False, False,  True,
       False,  True,  True])
>>> n*(c - lo)/(up - lo) == np.arange(n+1)
array([ True, False, False,  True, False, False,  True, False, False,
       False, False, False])
>>> n*((a - lo)/(up - lo)) == np.arange(n+1)
array([ True, False,  True, False, False,  True, False,  True,  True,
        True,  True,  True])
>>> n*((b - lo)/(up - lo)) == np.arange(n+1)
array([ True, False,  True, False, False,  True,  True, False,  True,
        True,  True,  True])
>>> n*((c - lo)/(up - lo)) == np.arange(n+1)
array([ True, False, False, False, False, False, False, False, False,
       False, False, False])
```
So I guess we cannot possibly do "exact" detection. Of the two indexing methods, the difference is smaller for `n*((x - lo)/(up - lo))`, and of the three binning methods, `a` and `b` are closer:
```python
>>> abs(n*(a - lo)/(up - lo) - np.arange(n+1)).max() / np.finfo("d").eps
8.0
>>> abs(n*(b - lo)/(up - lo) - np.arange(n+1)).max() / np.finfo("d").eps
8.0
>>> abs(n*(c - lo)/(up - lo) - np.arange(n+1)).max() / np.finfo("d").eps
16.0
>>> abs(n*((a - lo)/(up - lo)) - np.arange(n+1)).max() / np.finfo("d").eps
4.0
>>> abs(n*((b - lo)/(up - lo)) - np.arange(n+1)).max() / np.finfo("d").eps
4.0
>>> abs(n*((c - lo)/(up - lo)) - np.arange(n+1)).max() / np.finfo("d").eps
8.0
```
So perhaps those are the best to target, and we let the `is_close` tolerance be `4*eps`? i.e.
```cpp
std::abs(as_uniform.n * (edges[i] - as_uniform.lower) / (as_uniform.upper - as_uniform.lower) - (double) i) <= std::numeric_limits<double>::epsilon()*4
```
It's not clear if the factor of 4 is universal. If I set instead `n, lo, up = 53, 0., 500.` then the conclusions change a bit:
```python
>>> abs(n*(a - lo)/(up - lo) - np.arange(n+1)).max() / np.finfo("d").eps
32.0
>>> abs(n*(b - lo)/(up - lo) - np.arange(n+1)).max() / np.finfo("d").eps
32.0
>>> abs(n*(c - lo)/(up - lo) - np.arange(n+1)).max() / np.finfo("d").eps
32.0
>>> abs(n*((a - lo)/(up - lo)) - np.arange(n+1)).max() / np.finfo("d").eps
64.0
>>> abs(n*((b - lo)/(up - lo)) - np.arange(n+1)).max() / np.finfo("d").eps
32.0
>>> abs(n*((c - lo)/(up - lo)) - np.arange(n+1)).max() / np.finfo("d").eps
64.0
```
I'm not sure where we could comfortably go with this.

Reply:
Maybe the real answer is that the schema should evolve to allow specifying a uniform binning that is then interpreted specifically as one of the two indexing options shown?

Reply:
I agree we cannot use exact equality for floating points.

If "equal up to a certain precision" is not viable for correctionlib then I agree we can just drop this and just go for a separate type of correction, `UniformBinning`. I was thinking of proposing it in a future PR anyway for user ergonomics.

Reply:
An idea for the uniform binning is to extend the schema a tiny bit from https://github.com/cms-nanoAOD/correctionlib/blob/35039dbeeebe6f0731610d0abd82050e6a35fd25/src/correctionwlib/schemav2.py#L160-L162
to something like`edges: Union[List[float], UniformBinning]`
where `UniformBinning` is a dataclass with `start, stop, step` members. In the evaluator we can then handle it however we like, one option would be to just augment the Binning class to hold either one and dispatch as appropriate.